### PR TITLE
Add devices/properties badges

### PR DESCRIPTION
**Pull request by mthrok**
Add badges of supported properties and devices to functionals and transforms.

This commit adds `.. devices::` and `.. properties::` directives to sphinx.

APIs with these directives will have badges (based off of shields.io) which link to the 
page with description of these features.

Continuation of https://github.com/pytorch/audio/issues/2316
Excluded dtypes for further improvement, and actually added badges to most of functional/transforms.

### CODE REVIEW

File: docs/source/supported_features.rst [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+==================
+
+Each TorchAudio API supports different set of PyTorch features, such as different devices and data types.
+Supported features are indicated in API references like the following.
+
+.. devices:: CPU CUDA
+
+.. features:: Autograd TorchScript
+
+These icon mean that they are verified through automated testing.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:11]:
```suggestion
These icons mean that they are verified through automated testing.
```

File: docs/source/supported_features.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+.. features:: Autograd
+
+TorchAudio APIs with autograd support can correctly propagate the gradient in its backward path.
+
+For the basic of autograd, please checkout this `tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html>`_.
+
+.. note::
+
+   APIs without this mark might or might not raise and error in back propagation.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:87]:
```suggestion
   APIs without this mark may or may not raise and error in back propagation.
```

File: docs/source/supported_features.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+--------
+
+Autograd
+^^^^^^^^
+
+.. features:: Autograd
+
+TorchAudio APIs with autograd support can correctly propagate the gradient in its backward path.
+
+For the basic of autograd, please checkout this `tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html>`_.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:83]:
```suggestion
For the basics of autograd, please checkout this `tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html>`_.
```

File: docs/source/supported_features.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+   The lack of error in back propagatoin does not mean the feature computes the gradient correctly.
+
+TorchScript
+^^^^^^^^^^^
+
+.. features:: TorchScript
+
+TorchAudio APIs with TorchScript support can be serialized and executed on non-Python environments.
+
+For the deetail of TorchScript, please checkout the `documentation <https://pytorch.org/docs/stable/jit.html>`_.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:97]:
```suggestion
For the detail of TorchScript, please checkout the `documentation <https://pytorch.org/docs/stable/jit.html>`_.
```

File: docs/source/supported_features.rst [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+.. devices:: CPU CUDA
+
+.. features:: Autograd TorchScript
+
+These icon mean that they are verified through automated testing.
+
+.. note::
+
+   Missing feature icons mean that they are not tested, and this can mean
+   different things, depending on APIs.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:16]:
```suggestion
   different things, depending on the API.
```

File: docs/source/supported_features.rst [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+   1. The API is compatible with the feature but not tested.
+   2. The API is not compatible with the feature.
+
+   In case of 2, API might explicitly raise an error, but that is not always the case.
+   For example, APIs with missing Autograd badge might throw an error on backward path,
+   or silently returns a wrong gradient.
```
Comment (carolineechen) [inline docs/source/supported_features.rst:23]:
```suggestion
   or silently return a wrong gradient.
```

File: docs/source/custom_directives.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        raise ValueError(
+            f"One or more device values are not valid. The valid values are {valid_values}. Given value: '{arg}'"
+        )
+    return ", ".join(sorted(devices))
+
+
+def _parse_feature_args(arg: str):
+    features = sorted(arg.strip().split())
+
+    valid_values = {"Autograd", "TorchScript", "Kaldi-Compatible"}
```
Comment (carolineechen) [inline docs/source/custom_directives.py:71]: Kaldi-Compatible is missing from the supported features documentation and also do not see it being used anywhere -- is it going to be added in a separate PR?

File: docs/source/supported_features.rst [hunk 8]
```diff
@@ -0,0 +1,97 @@
+Supported Features
```
Comment (carolineechen) [inline docs/source/supported_features.rst:1]: thoughts on renaming this "Supported Properties" or something of that nature? could be slightly confusing to use feature as both the overall feature set and as a subset (autograd/torchscript)

File: torchaudio/functional/functional.py [hunk 9]
```diff
@@ -1429,6 +1505,10 @@ def resample(
 ) -> Tensor:
     r"""Resamples the waveform at the new frequency using bandlimited interpolation.
 
+    .. devices:: CPU CUDA
```
Comment (carolineechen) [inline torchaudio/functional/functional.py:1508]: could you move this down under the link below?

File: torchaudio/functional/functional.py [hunk 10]
```diff
@@ -464,6 +484,8 @@ def melscale_fbanks(
 ) -> Tensor:
     r"""Create a frequency bin conversion matrix.
 
+    .. devices:: CPU
+
```
Comment (carolineechen) [inline torchaudio/functional/functional.py:488]: `melscale_fbanks`, `linear_fbanks`, and `create_dct` are all torchscript-tested as well

File: docs/source/supported_features.rst (hunk 8 above)
Comment (mthrok) [reply docs/source/supported_features.rst:1]: Yes! I was also feeling that "feature" is overloaded here. Thanks for the suggestion. Will do.

File: docs/source/custom_directives.py (hunk 7 above)
Comment (mthrok) [reply docs/source/custom_directives.py:71]: Oh, it was some idea I got in the middle. I will remove it for now.

File: torchaudio/functional/functional.py (hunk 9 above)
Comment (mthrok) [reply torchaudio/functional/functional.py:1508]: [PEP257](https://peps.python.org/pep-0257/#multi-line-docstrings
) recommends multi-line-docstrings to be split into one-line summary and elaborated explanation.

I would like to put badges between the summary and the rest, and the URL does not fit nicely.

Can I add some more explanation along the URL?

Comment (mthrok) [reply torchaudio/functional/functional.py:1514]: The homepage has an example to cite the page. https://ccrma.stanford.edu/~jos/resample/resample-citation.html

I moved it to bibtex reference. How does it look?

File: docs/source/supported_features.rst (hunk 8 above)
Comment (mthrok) [reply docs/source/supported_features.rst:1]: Updated. It still uses the ambiguous `feature` term. Let me know what you think.

File: docs/source/supported_features.rst [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+.. properties:: Autograd
+
+TorchAudio APIs with autograd support can correctly propagate the gradient in its backward path.
+
+For the basics of autograd, please checkout this `tutorial <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html>`_.
+
+.. note::
+
+   APIs without this mark may or may not raise and error in back propagation.
```
Comment (hwangjeff) [inline docs/source/supported_features.rst:88]:
```suggestion
   APIs without this mark may or may not raise an error during backpropagation.
```

File: docs/source/supported_features.rst [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+   The lack of error in back propagatoin does not mean the gradient is correct.
```
Comment (hwangjeff) [inline docs/source/supported_features.rst:89]:
```suggestion
   The absence of an error raised during backpropagation does not necessarily mean the gradient is correct.
```

File: docs/source/supported_features.rst [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+TorchScript
+^^^^^^^^^^^
+
+.. properties:: TorchScript
+
+TorchAudio APIs with TorchScript support can be serialized and executed on non-Python environments.
```
Comment (hwangjeff) [inline docs/source/supported_features.rst:96]:
```suggestion
TorchAudio APIs with TorchScript support can be serialized and executed in non-Python environments.
```

File: docs/source/supported_features.rst [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+For the detail of TorchScript, please checkout the `documentation <https://pytorch.org/docs/stable/jit.html>`_.
```
Comment (hwangjeff) [inline docs/source/supported_features.rst:98]:
```suggestion
For details on TorchScript, please refer to the `documentation <https://pytorch.org/docs/stable/jit.html>`_.
```

**Review by hwangjeff — APPROVED** [review]
just cosmetic things — looks good overall

File: torchaudio/functional/functional.py (hunk 9 above)
Comment (carolineechen) [reply torchaudio/functional/functional.py:1514]: yup that looks good

File: docs/source/supported_features.rst (hunk 8 above)
Comment (mthrok) [reply docs/source/supported_features.rst:1]: I am thinking to mention numerical compatibility with other libraries such as librosa and Kaldi, so let me follow-up on this one.
